File: opt/poly_fit.py
"""
poly_fit.py - General Purpose Polynomial Curve Fitting with Error Analysis

Fit a power-polynomial y_fit(x;a) to data pairs (x,y) where:
    y_fit(x;a) = SUM_i a_i * x^p_i

Minimizes the Chi-square error criterion with optional regularization.
Provides comprehensive error analysis including parameter uncertainties,
confidence intervals, correlation matrix, and information criteria.

Translation from MATLAB by Claude, 2025-10-24
Original by H.P. Gavin

Reference:
H.P. Gavin, "Fitting Models to Data: Generalized Linear Least Squares 
and Error Analysis"
https://people.duke.edu/~hpgavin/SystemID/linear-least-sqaures.pdf
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm


def poly_fit(x, y, p, figNo=0, Sy=None, rof=None, b=0.0):
    """
    Fit a power-polynomial to data with comprehensive error analysis.
    
    Fits: y_fit(x;c) = SUM_i c_i * x^p_i
    
    Minimizes Chi-square criterion:
        X2 = SUM_k [(y_fit(x_k;c) - y_k)^2 / Sy_k^2]
    
    Parameters
    ----------
    x : array_like, shape (m,)
        Known vector of independent variables
    y : array_like, shape (m,)
        Measured vector of dependent variables
    p : array_like, shape (n,)
        Vector of real powers (x^p) for each polynomial term
    figNo : int, optional
        Figure number for plotting. Use 0 to suppress plotting (default: 0)
    Sy : float or array_like, optional
        Measurement errors for each value of y. 
        Scalar or shape (m,) (default: 1.0)
    rof : array_like, shape (2,), optional
        Range of fit [x_min, x_max] (default: [min(x), max(x)])
    b : float, optional
        Regularization constant (default: 0.0)
    
    Returns
    -------
    c : ndarray, shape (n,)
        Values of polynomial coefficients
    x_fit : ndarray, shape (100,)
        Values of x within range of fit
    y_fit : ndarray, shape (100,)
        Polynomial evaluated at x_fit
    Sc : ndarray, shape (n,)
        Standard errors of polynomial coefficients
    Sy_fit : ndarray, shape (100,)
        Standard errors of the curve fit
    Rc : ndarray, shape (n, n)
        Parameter correlation matrix
    R2 : float
        R-squared error criterion
    Vr : float
        Unbiased variance of unweighted residuals
    AIC : float
        Akaike Information Criterion
    condNo : float
        Condition number of regularized system matrix
    
    Notes
    -----
    Unlike numpy.polyfit, this function allows:
    - Any real-valued powers (not just integer exponents)
    - Weighted least squares with measurement errors
    - Regularization parameter
    - Comprehensive error analysis and visualization
    """
    
    # Convert inputs to numpy arrays
    x = np.asarray(x).flatten()
    y = np.asarray(y).flatten()
    p = np.asarray(p).flatten()
    
    # Error checking
    if len(x) != len(y):
        raise ValueError('Length of x must equal length of y')
    
    Nd = len(x)  # Number of data points
    Np = len(p)  # Number of parameters
    Nf = 100     # Number of values in the fit
    
    # Default range of fit (with 5% extrapolation)
    ee = 0.05
    rfd = np.array([min(x), max(x)]) @ np.array([[1+ee, -ee], [-ee, 1+ee]])
    
    # Handle optional arguments
    if Sy is None:
        Sy = 1.0
        compute_Vr = True  # Will compute measurement errors from residuals
    else:
        compute_Vr = False
        Sy = np.asarray(Sy)
        if np.any(Sy <= 0):
            Sy = np.where(Sy <= 0, 1.0, Sy)
    
    if rof is None:
        rof = rfd
    
    # Ensure Sy is proper shape
    if np.isscalar(Sy):
        Sy = Sy * np.ones(Nd)
    
    # x values for the fit
    x_fit = np.linspace(rof[0], rof[1], Nf)
    
    # Inverse measurement error covariance matrix
    ISy = np.diag(1.0 / (Sy**2))
    
    # Build design matrices
    B = np.zeros((Nd, Np))
    B_fit = np.zeros((Nf, Np))
    
    for i in range(Np):
        B[:, i] = x ** p[i]          # Model basis vectors for data
        B_fit[:, i] = x_fit ** p[i]  # Model basis vectors for fit
    
    # Condition number
    condNo = np.linalg.cond(B.T @ ISy @ B + b * np.eye(Np))
    
    # Least squares parameters
    c = np.linalg.solve(B.T @ ISy @ B + b * np.eye(Np), B.T @ ISy @ y)
    
    # Least squares fit
    y_fit = B_fit @ c
    
    # Unbiased variance of unweighted residuals
    Vr = np.sum((y - B @ c)**2) / (Nd - Np - 1)
    
    # Measurement error covariance
    if compute_Vr:
        invVy = np.eye(Nd) / Vr  # Computed from residuals
    else:
        invVy = ISy              # Provided by user
    
    # Parameter covariance matrix
    Vc = np.linalg.inv(B.T @ invVy @ B + b * np.eye(Np))
    
    # Regularized least squares adjustment
    if b != 0:
        Vy = np.linalg.inv(ISy)
        Vc = Vc @ (B.T @ ISy @ Vy @ ISy @ B) @ Vc
    
    # Standard errors of parameters
    Sc = np.sqrt(np.diag(Vc))
    
    # Parameter cross-correlation matrix
    Rc = Vc / np.outer(Sc, Sc)
    
    # Standard error of the fit
    Sy_fit = np.sqrt(np.diag(B_fit @ Vc @ B_fit.T))
    
    # R-squared (coefficient of determination)
    R2 = 1 - np.sum((y - B @ c)**2) / np.sum((y - np.mean(y))**2)
    
    # Akaike Information Criterion
    AIC = np.log(2 * np.pi * Np * Vr) + (B @ c - y).T @ invVy @ (B @ c - y) + 2 * Np
    
    # Print results
    print('\n' + '='*70)
    print('Polynomial Fit Results')
    print('='*70)
    print('     p         c            +/-   dc           (percent)')
    print('-'*65)
    for i in range(Np):
        pct = 100 * Sc[i] / abs(c[i]) if c[i] != 0 else np.inf
        if p[i] == int(p[i]):
            print(f'   c[{int(p[i]):2d}] =  {c[i]:11.3e}    +/- {Sc[i]:10.3e}    '
                  f'({pct:7.2f} %)')
        else:
            print(f' {p[i]:8.2f} :  {c[i]:11.3e}     +/- {Sc[i]:10.3e}    '
                  f'({pct:7.2f} %)')
    print('='*70 + '\n')
    
    # Plotting
    if figNo > 0:
        _plot_results(x, y, x_fit, y_fit, B, c, Sy_fit, Vr, 
                     condNo, R2, AIC, Nd, Np, figNo)
    
    return c, x_fit, y_fit, Sc, Sy_fit, Rc, R2, Vr, AIC, condNo


def _plot_results(x, y, x_fit, y_fit, B, c, Sy_fit, Vr, 
                  condNo, R2, AIC, Nd, Np, figNo):
    """
    Create visualization of polynomial fit results.
    
    Internal function called by poly_fit when figNo > 0.
    """
    
    # Confidence intervals
    CI = np.array([0.90, 0.99])
    z = norm.ppf(1 - (1 - CI) / 2)
    
    # Confidence bands for the model
    yps95 = y_fit + z[0] * Sy_fit
    yms95 = y_fit - z[0] * Sy_fit
    yps99 = y_fit + z[1] * Sy_fit
    yms99 = y_fit - z[1] * Sy_fit
    
    # Coordinates for patch plots
    xp = np.concatenate([x_fit, x_fit[::-1], [x_fit[0]]])
    yp95 = np.concatenate([yps95, yms95[::-1], [yps95[0]]])
    yp99 = np.concatenate([yps99, yms99[::-1], [yps99[0]]])
    
    # Colors for confidence intervals
    patchColor95 = [0.95, 0.95, 0.1]
    patchColor99 = [0.2, 0.95, 0.2]
    
    # Set plotting style
    plt.rcParams.update({
        'font.size': 12,
        'lines.linewidth': 2,
        'lines.markersize': 5
    })
    
    # Figure 1: Data, fit, and confidence intervals
    fig = plt.figure(figNo, figsize=(14, 6))
    fig.clf()
    
    # Left subplot: Data and model with confidence intervals
    ax1 = plt.subplot(1, 2, 1)
    ax1.fill(xp, yp99, color=patchColor99, edgecolor=patchColor99, 
             alpha=0.3, label=f'{int(CI[1]*100)}% c.i.')
    ax1.fill(xp, yp95, color=patchColor95, edgecolor=patchColor95, 
             alpha=0.5, label=f'{int(CI[0]*100)}% c.i.')
    ax1.plot(x, y, 'ob', linewidth=3, markersize=6, label='data')
    ax1.plot(x_fit, y_fit, '-k', linewidth=2, label='y_fit')
    ax1.set_xlabel('x', fontsize=13)
    ax1.set_ylabel('y', fontsize=13)
    ax1.legend(loc='best', fontsize=11)
    ax1.set_xlim([min(xp), max(xp)])
    y_range = max(y) - min(y)
    ax1.set_ylim([min(y) - 0.1*y_range, max(y) + 0.1*y_range])
    ax1.grid(True, alpha=0.3)
    ax1.set_title('Polynomial Fit with Confidence Intervals', fontsize=14)
    
    # Right subplot: Data vs model (correlation plot)
    ax2 = plt.subplot(1, 2, 2)
    ax2.plot(y, y, '-k', linewidth=0.5, alpha=0.5)  # 1-to-1 line
    ax2.plot(B @ c, y, 'ob', linewidth=3, markersize=6)
    
    # Add statistics text
    tx = 0.90 * min(y) + 0.10 * max(y)
    ty_range = max(y) - min(y)
    positions = [0.98, 0.90, 0.82, 0.74, 0.66]
    
    ax2.text(tx, min(y) + positions[0]*ty_range, 
             f'cond # = {condNo:.1f}', fontsize=11)
    ax2.text(tx, min(y) + positions[1]*ty_range, 
             f'σ_r = {np.sqrt(Vr):.3f}', fontsize=11)
    ax2.text(tx, min(y) + positions[2]*ty_range, 
             f'AIC = {AIC:.1f}', fontsize=11)
    ax2.text(tx, min(y) + positions[3]*ty_range, 
             f'R² = {R2:.3f}', fontsize=11)
    ax2.text(tx, min(y) + positions[4]*ty_range, 
             f'n = {Np}', fontsize=11)
    
    ax2.set_xlabel('y_fit', fontsize=13)
    ax2.set_ylabel('y', fontsize=13)
    ax2.axis('tight')
    ax2.grid(True, alpha=0.3)
    ax2.set_title('Model vs Data (Correlation)', fontsize=14)
    
    plt.tight_layout()
    
    # Figure 2: Histogram of residuals
    residuals = y - B @ c
    nBars = max(10, round(Nd / 5))
    
    fig2 = plt.figure(figNo + 1, figsize=(10, 6))
    fig2.clf()
    
    counts, bins, _ = plt.hist(residuals, bins=nBars, 
                                color=[0.3, 0.7, 0.9], 
                                edgecolor='black', alpha=0.7)
    plt.xlabel('Residuals, r = y - y_fit', fontsize=13)
    plt.ylabel('Empirical PDF, f_R(r)', fontsize=13)
    plt.title('Distribution of Residuals', fontsize=14)
    plt.grid(True, alpha=0.3, axis='y')
    
    # Add normal distribution overlay
    mu, std = np.mean(residuals), np.std(residuals)
    xmin, xmax = plt.xlim()
    x_normal = np.linspace(xmin, xmax, 100)
    p_normal = norm.pdf(x_normal, mu, std)
    # Scale to match histogram
    p_normal_scaled = p_normal * len(residuals) * (bins[1] - bins[0])
    plt.plot(x_normal, p_normal_scaled, 'r-', linewidth=2, 
             label=f'Normal(μ={mu:.3f}, σ={std:.3f})')
    plt.legend(fontsize=11)
    
    plt.tight_layout()
    
    # Figure 3: CDF of residuals with confidence intervals
    plot_CDF_ci(residuals, 95, figNo + 2)


def plot_CDF_ci(data, confidence_level, figNo):
    """
    Plot empirical CDF of data with confidence intervals.
    
    Parameters
    ----------
    data : array_like
        Data values
    confidence_level : float
        Confidence level (e.g., 95 for 95%)
    figNo : int
        Figure number
    """
    
    data = np.asarray(data).flatten()
    n = len(data)
    
    # Sort data
    sorted_data = np.sort(data)
    
    # Empirical CDF
    ecdf = np.arange(1, n + 1) / n
    
    # Confidence intervals using Dvoretzky-Kiefer-Wolfowitz inequality
    alpha = 1 - confidence_level / 100
    epsilon = np.sqrt(np.log(2 / alpha) / (2 * n))
    
    upper_ci = np.minimum(ecdf + epsilon, 1.0)
    lower_ci = np.maximum(ecdf - epsilon, 0.0)
    
    # Theoretical normal CDF
    mu, std = np.mean(data), np.std(data)
    x_theory = np.linspace(min(sorted_data), max(sorted_data), 200)
    cdf_theory = norm.cdf(x_theory, mu, std)
    
    # Plotting
    fig = plt.figure(figNo, figsize=(10, 7))
    fig.clf()
    
    plt.fill_between(sorted_data, lower_ci, upper_ci, 
                     color=[0.8, 0.9, 1.0], alpha=0.5,
                     label=f'{confidence_level}% confidence band')
    plt.plot(sorted_data, ecdf, 'o-', color=[0.2, 0.4, 0.8], 
             linewidth=2, markersize=4, label='Empirical CDF')
    plt.plot(x_theory, cdf_theory, 'r-', linewidth=2, 
             label=f'Normal CDF (μ={mu:.3f}, σ={std:.3f})')
    plt.plot(sorted_data, sorted_data * 0 + 0.5, '--k', 
             alpha=0.3, linewidth=1)
    
    plt.xlabel('Residual value', fontsize=13)
    plt.ylabel('Cumulative probability', fontsize=13)
    plt.title(f'Empirical CDF with {confidence_level}% Confidence Intervals', 
              fontsize=14)
    plt.legend(fontsize=11)
    plt.grid(True, alpha=0.3)
    plt.ylim([0, 1])
    
    plt.tight_layout()


# Test and demonstration
if __name__ == '__main__':
    """
    Test poly_fit with example data
    """
    
    print("\n" + "="*70)
    print("Testing poly_fit.py")
    print("="*70)
    
    # Generate test data
    np.random.seed(42)
    x_l = -1
    x_h = 1
    Nd = 40
    
    x = np.linspace(x_l, x_h, Nd)
    
    measurement_error = 0.20
    y = -np.cos(4*x) + 1.0 * x**3 * np.exp(-x/3) + measurement_error * np.random.randn(Nd)
    
    print(f"\nGenerated {Nd} data points")
    print(f"x range: [{x_l}, {x_h}]")
    print(f"Measurement error (RMS): {measurement_error}")
    
    # Test 1: Full polynomial basis
    print("\n" + "-"*70)
    print("Test 1: Powers [0, 1, 2, 3, 4]")
    print("-"*70)
    
    p1 = np.array([0, 1, 2, 3, 4])
    c1, x_fit1, y_fit1, Sc1, Sy_fit1, Rc1, R2_1, Vr1, AIC1, cond1 = \
        poly_fit(x, y, p1, figNo=10)
    
    # Test 2: Reduced polynomial basis (no linear term)
    print("\n" + "-"*70)
    print("Test 2: Powers [0, 2, 3, 4]")
    print("-"*70)
    
    p2 = np.array([0, 2, 3, 4])
    c2, x_fit2, y_fit2, Sc2, Sy_fit2, Rc2, R2_2, Vr2, AIC2, cond2 = \
        poly_fit(x, y, p2, figNo=20)
    
    # Comparison
    print("\n" + "="*70)
    print("Comparison of Models")
    print("="*70)
    print(f"{'Metric':<20} {'Model 1 (5 terms)':<20} {'Model 2 (4 terms)':<20}")
    print("-"*70)
    print(f"{'R²':<20} {R2_1:<20.4f} {R2_2:<20.4f}")
    print(f"{'AIC':<20} {AIC1:<20.2f} {AIC2:<20.2f}")
    print(f"{'Cond. Number':<20} {cond1:<20.1f} {cond2:<20.1f}")
    print(f"{'Residual Var':<20} {Vr1:<20.4f} {Vr2:<20.4f}")
    print("="*70)
    
    if AIC2 < AIC1:
        print("\n✓ Model 2 is preferred (lower AIC)")
    else:
        print("\n✓ Model 1 is preferred (lower AIC)")
    
    plt.show()
    
    print("\n" + "="*70)
    print("poly_fit test completed successfully!")
    print("="*70)

File: src/fit/m-files/poly_fit.m
function [c,x_fit,y_fit,Sc,Sy_fit,Rc,R2,Vr,AIC,condNo] = poly_fit(x,y,p,figNo,Sy,rof,b)
% [c,x_fit,y_fit,Sc,Sy_fit,Rc,R2,Vr,AIC,condNo] = poly_fit(x,y,p,figNo,Sy,rof,b)
%
% fit a power-polynomial, y_fit(x;a) to data pairs (x,y) where
%
%  y_fit(x;a) = SUM_i=1^length(p)  c_i  x^p_i
%
% which minimizes the Chi-square error criterion, X2,
%
%  X2 = SUM_k=1 ^ length(x) [ ( y_fit(x_k;a) - y_k )^2 / Sy_k^2 ]
%
% where  Sy_k is the standard error of the k-th data point.  
%
%  INPUT     DESCRIPTION                                        DIM     DEFAULT
% ------     ------------------------------------------------  -----    -------
%   x        known vector of the independent variables,        (1xm)
%   y        measured vector of the dependent variables,       (1xm)
%   p        vector of the real power of x (x^p) of each term  (1xn)
%   figNo    figure number. Use 0 to suppress plotting                    0
%   Sy       measurement errors for each value of y  (1x1) or  (1xm)      1
%   rof      range of the fit                                  (1x2) [xMin,xMax]
%   b        regularization constant                                      0
%
% OUTPUT     DESCRIPTION                                         
% ------     ----------------------------------------------------
%   c        values of the polynomial coefficients
%   x_fit    values of x within the range of fit fit 
%   y_fit    poloynomial at values of x_fit
%   Sc       standard errors of polynomial coefficients 
%   Sy_fit   standard errors of the curve-fit
%   Rc       parameter correlation matrix
%   R2       R-squared error criterion
%   Vr       unbiased variance of unweighted residuals
%   AIC      Akaike information criterion
%   condNo   condition number of the regularized system matrix 
%
% HP Gavin, 
% Fitting Models to Data: Generalized Linear Least Squares and Error Analysis
% https://people.duke.edu/~hpgavin/SystemID/linear-least-sqaures.pdf
%
% updated 2007-04-10, 2013-02-24, 2024-11-04

 if ( length(x) ~= length(y) )       % error checking
   disp(' the length of x must equal the length of y ');
   return
 end

 ee = 0.05;
 rfd = [ min(x) , max(x) ]*[ 1+ee -ee ; -ee 1+ee ]; % extrapolated range of fit
 if nargin < 4,  figNo = 0;  end       % default figure number
 if nargin < 5,  Sy  = 1.0;  end       % default measurement error
 if nargin < 6,  rof = rfd;  end       % default range of the fit 
 if nargin < 7,  b   = 0.0;  end       % default regularization parameter

 if any(Sy <= 0)                       % ensure that all vaues of Sy are positve
   Sy(find(Sy<=0)) = 1.0;
%  nargin = 4;                         % use Vr to compute invVy, Vc, and Sy_fit
 end

 x = x(:);                             % make "x" a column-vector
 y = y(:);                             % make "y" a column-vector
 p = p(:);                             % make "p" a column-vector

 Nd = length(x);                       % number of data points
 Np = length(p);                       % number of parameters
 Nf = 100;                             % number of values in the fit 
 x_fit = linspace(rof(1),rof(2),Nf)';  % x values for the fit

 ISy = eye(Nd)./(Sy.^2);       % inverse of the measuremen error covariance, Vy

 B     = zeros(Nd,Np);         % allocate memory for the model basis vectors
 B_fit = zeros(Nf,Np);         % allocate memory for the model basis vectors

 for i=1:Np
   B (:,i) = x.^p(i);          % the matrix of model basis vectors 
   B_fit(:,i) = x_fit.^p(i);   % the matrix of model basis vectors 
 end

 condNo = cond( B'*ISy*B + b*eye(Np) );              % condition number 

 c = inv( B'*ISy*B + b*eye(Np) ) * B'*ISy*y;         % least squares parameters

 y_fit = B_fit*c;                                    % least squares fit

 Vr = sum((y-B*c).^2) / (Nd-Np-1);  % unbiased variance of unweighted residuals

 if nargin < 5
   invVy = eye(Nd) / Vr;  % measurement error covariance computed from residuals
 else 
   invVy = ISy;           % measurement error covariance provided by user
 end

 Vc = inv( B'*invVy*B + b*eye(Np) );    % estimated parameter covariance

 if b ~= 0                              % regularized least squares 
   Vc = Vc * (B'*ISy*Vy*ISy*B) * Vc;    % parameter covariance w/ regularization
 end

 Sc = sqrt(diag(Vc));                   % standard error of the parameters

 Rc = Vc ./ (Sc * Sc');                 % parameter cross-correlation matrix

 Sy_fit = sqrt(diag(B_fit*Vc*B_fit'));  % standard error of the fit 

 R2  = 1 - sum( (y-B*c).^2 ) / sum( (y-sum(y)/Nd).^2 ); % R-squared = correlation

 AIC =  log(2*pi*Np*Vr) + (B*c-y)'*invVy*(B*c-y) + 2*Np; 

 disp('     p         a            +/-   da           (percent) ')
 disp('---------------------------------------------------------')
 for i=1:Np
   if rem(p,1) == 0
      fprintf('   a[%2d] =  %11.3e;    +/- %10.3e    (%7.2f %%)\n',  ...
                                p(i), c(i), Sc(i), 100*Sc(i)/abs(c(i)) ); 
   else
      fprintf(' %8.2f :  %11.3e     +/- %10.3e    (%7.2f %%)\n',  ...
                                p(i), c(i), Sc(i), 100*Sc(i)/abs(c(i)) ); 
   end
 end

if figNo % Plots ---------------------------------------
 
  formatPlot(12,2, 5);  % font size, line width, marker size

  figure(figNo);        % plot the datc, the model, and the confidence interval
    
   CI = [ 0.90  0.99 ];    % confidence intervals 
   z  = norminv( 1 - (1 - CI) / 2 );    % z-scores corresponding to confidence intervals

    % confidence interval (CI) of the model (y_fit)
    yps95 =  y_fit + z(1)*(Sy_fit + 0*sqrt(Vr));       %  + 90 CI
    yms95 =  y_fit - z(1)*(Sy_fit + 0*sqrt(Vr));       %  - 90 CI
    yps99 =  y_fit + z(2)*(Sy_fit + 0*sqrt(Vr));       %  + 99 CI
    yms99 =  y_fit - z(2)*(Sy_fit + 0*sqrt(Vr));       %  - 99 CI
    xp  = [ x_fit ; x_fit(end:-1:1) ; x_fit(1) ];    % x coordinates for patch
    yp95  = [ yps95 ; yms95(end:-1:1) ; yps95(1) ];  % y coordinates for patch
    yp99  = [ yps99 ; yms99(end:-1:1) ; yps99(1) ];  % y coordinates for patch

    patchColor95 = [ 0.95, 0.95, 0.1 ];              % color for the 95% CI
    patchColor99 = [ 0.2,  0.95, 0.2 ];              % color for the 99% CI

    clf
    subplot(121)        % plot the datc, curvefit, and CI's
    hold on
      hc99 = patch( xp, yp99, patchColor99);         % 99% CI
      hc95 = patch( xp, yp95, patchColor95);         % 95% CI
      hd   = plot( x, y,     'ob',  'LineWidth', 3); % data points
      hf   = plot( x_fit, y_fit, '-k');              % polynomial fit
    hold off
    set(hc95,'EdgeColor', patchColor95 );
    set(hc99,'EdgeColor', patchColor99 );
    xlabel('x') 
    ylabel('y')
    legend([hd, hf , hc95, hc99 ], 'data', 'y_{fit}', sprintf('%2.0f%% c.i.',CI(1)*100), sprintf('%2.0d%% c.i.',CI(2)*100) );
    axis([ min(xp) , max(xp) , min(y)-0.1*(max(y)-min(y)) , max(y)+0.1*(max(y)-min(y)) ])

% {
    subplot(122)        % plot the data vs the model and show correlation 
    hold on 
      plot( y, y,'-k', 'LineWidth', 0.5 )             % 1-to-1 line
      plot( B*c, y, 'ob', 'LineWidth', 3)             % data and model points
      tx = 0.90*min(y) + 0.10*max(y);
      ty = 0.02*min(y) + 0.98*max(y); 
      ry  = [ max(y) ; min(y) ];
      text(tx,[1.0,0.0]*ry, sprintf('condn # = %4.1f', condNo));
      text(tx,[0.9,0.1]*ry, sprintf('\\sigma_r = %5.3f', sqrt(Vr)));
      text(tx,[0.8,0.2]*ry, sprintf('AIC = %4.1f', AIC));
      text(tx,[0.7,0.3]*ry, sprintf('R^2 = %5.3f', R2));
      text(tx,[0.6,0.4]*ry, sprintf('n = %d', length(c) ));
    hold off
    xlabel('y_{fit}')
    ylabel('y')
    axis('tight')
% }

  figure(figNo+1)          % plot the histogram (distribution) of the residuals
    clf
    nBars = max(10,round(Nd/5)); 
    [fx,xx] = hist(y-B*c, nBars, nBars/(max(x)-min(x)));
      bar(xx,fx)        % histogram of the residuals
      xlabel('residuals, r = y - y_{fit}')
      ylabel('empirical PDF, f_R(r)')

  plotCDFci(y-B*c, 95, figNo+2);  % CDF of residuals with confidence intervals

end % ------------------------------------------------------- Plots

% ------------------------------------------------------------------ poly_fit
